README
NAME
    Net::TiVo - Perl interface to TiVo.

INSTALLATION
    % perl Makefile.PL 
    % make
    % make test
    % make install

REQUIREMENTS
    XML::Simple
    Log::Log4perl
    LWP::UserAgent

PLATFORMS
    I have tested it on the following platforms, but it should work on any platform
    that runs Perl, and meets the above requirements.
    
    FreeBSD v6.1

SYNOPSIS
        use Net::TiVo;

        my $tivo = Net::TiVo->new(
            host => '192.168.1.25', 
            mac  => 'MEDIA_ACCESS_KEY'
        );

        for ($tivo->folders()) {
            print $_->as_string(), "\n";
        }   

ABSTRACT
    "Net::TiVo" provides an object-oriented interface to TiVo's REST
    interface. This makes it possible to enumerate the folders and shows,
    and dump their meta-data.

DESCRIPTION
    "Net::TiVo" has a very simple interface, and currently only supports the
    enumeration of folder and shows using the REST interface. The main
    purpose of this module was to provide access to the TiVo
    programmatically to automate the process of downloading shows from a
    TiVo.

    "Net::TiVo" does not provide support for downloading from TiVo. There
    are several options available, including LWP, wget, and curl. Note: I
    have used wget version >= 1.10 with success. wget version 1.09 appeared
    to have an issue with TiVo's cookie.

CACHING
    "Net::TiVo" is slow due to the amount of time it takes to fetch data
    from TiVo. This is greatly sped up by using a cache. "Net::TiVo"'s "new"
    method accepts a reference to a "Cache" object. Any type of caching
    object may be supported as long as it meets the requirements below.
    There are several cache implementations available on CPAN, such as
    "Cache::Cache".

    The following example creates a cache that lasts for 600 seconds.

        use Cache::FileCache;
    
        my $cache = Cache::FileCache->new(
             namespace          => 'TiVo',
             default_expires_in => 600,
        }

        my $tivo = Net::TiVo->new(
             host  => '192.168.1.25',
             mac   => 'MEDIA_ACCESS_KEY',
             cache => $cache,
        }

    "Net::TiVo" uses *positive* caching, errors are not stored in the cache.

    Any "Cache" class may be used as long as it supports the following
    method signatures.

        # Set a cache value
        $cache->set($key, $value);

        # Get a cache value
        $cache->get($key);

  METHODS
    folders()
        Returns an array in list context or array reference in scalar
        context containing a list of Net::TiVo::Folder objects.

SEE ALSO
    Net::TiVo::Folder, Net::TiVo::Show

AUTHOR
    Christopher Boumenot, <boumenot@gmail.com>



Changes
0.10 (12/10/2007)
    (cmb) David Steele reported that Net::TiVo does not depend on 
          Crypt::SSLeay, and it should.

0.09 (10/27/2007)
    (cmb) Artem Belevich emailed me with another MIME type for his TiVo 
          HD.

0.08 (10/10/2007)
    (cmb) Michael Lindner was getting 500 errors, and eventually tracked
          it down to LWP and Net::SSLeay.  Once he switched to 
          Crypt::SSLeay the 500 errors went away.
    (cmb) David Steele reported an issue with an empty suggestions folder
          that caused Net-TiVo to die unnecessarily.
    (cmb) David Steele reported the the MIME has changed from
          video/x-tivo-mpeg to video/x-tivo-raw-pes.
    (cmb) Add the method high_definition for the Show object.
    (cmb) Add canned data for the new schema, and update the shows test.

0.07 (01/12/2007)
    (cmb) Steven Schubiger clarified the documentation for folders().
          Added his suggestion about using L<> for the SEE ALSO section.
    (cmb) Added documentation about downloading shows from TiVo using 
          LWP, wget, or curl.
    (cmb) Added documentation about using a cache.
    (cmb) Updated the README, and changed the format slightly of Changes.

0.06 (01/10/2007)
    (cmb) David Steele noticed a problem when using the example in the 
          documentation.  The documentation was wrong, it has now been
          fixed
    (cmb) Fix the wording, and grammar for some of the documentation, 
          and comments.
    (cmb) Update the copyright of the library.

0.05 (12/30/2006)
    (cmb) Patch by Kevin Brintnall to clarify the return of the 
          folders function.

0.04 (12/29/2006)
    (cmb) Patch by Richard G. Harman Jr., skip folders that have no 
          shows.

0.03 (08/02/2006)
    (cmb) Ahhh!  I wasn't including Makefile.PL in the distribution.

0.02 (07/29/2006)
    (cmb) Fix the POD

0.01 (07/29/2006)
    (cmb) Initial release


lib/Net/TiVo/Show.pm
# $Id: Show.pm 57 2007-01-12 19:26:09Z boumenot $
# Author: Christopher Boumenot <boumenot@gmail.com>
######################################################################
#
# Copyright 2006-2007 by Christopher Boumenot.  This program is free 
# software; you can redistribute it and/or modify it under the same
# terms as Perl itself.
#
######################################################################

package Net::TiVo::Show;

use strict;
use warnings;
use base qw(Net::TiVo::Folder);

use Text::Wrap;

# Should be read as a poor man's XPath
our %DEFAULT_ATTRIBUTES_XPATH = (
    station      => [qw(Details SourceStation)],
    name         => [qw(Details Title)],
    episode      => [qw(Details EpisodeTitle)],
    episode_num  => [qw(Details EpisodeNumber)],                                 
    content_type => [qw(Details ContentType)],
    capture_date => [qw(Details CaptureDate)],    
    format       => [qw(Details SourceFormat)],
    high_definition => [qw(Details HighDefinition)],
    size         => [qw(Details SourceSize)],
    channel      => [qw(Details SourceChannel)],
    duration     => [qw(Details Duration)],
    description  => [qw(Details Description)],
    series_id    => [qw(Details SeriesId)],
    program_id   => [qw(Details ProgramId)],
    url          => [qw(Links Content Url)],
);

__PACKAGE__->make_accessor($_) for keys %DEFAULT_ATTRIBUTES_XPATH;
__PACKAGE__->make_accessor($_) for qw(tuner);

sub new {
    my ($class, %options) = @_;
    
    unless ($options{xmlref}) {
        die __PACKAGE__ . ": Mandatory param xmlref missing\n";
    }

    my $self = {
        %options,
    };

    bless $self, $class;

    for my $attr (keys %DEFAULT_ATTRIBUTES_XPATH) {
        my $value = __PACKAGE__->walk_hash_ref($options{xmlref}, $DEFAULT_ATTRIBUTES_XPATH{$attr});
        $self->$attr($value);
    }

    # do a little post processing 
    $self->capture_date(hex($self->capture_date()));

    my ($channel, $tuner) = split(/\-/, $self->channel());
    $tuner = 0 unless defined $tuner;
    $self->channel($channel);
    $self->tuner($tuner);

    return $self;
}

sub as_string {
    my $self = shift;

    $Text::Wrap::columns = 72;

    my @a;
    push @a, $self->name();
    push @a, $self->episode();
    push @a, $self->description();
    push @a, int(($self->duration() / (60 * 1000)) + 0.5) . " min";

    my $s = wrap("", "      ", join(", ", @a));
    $s .= "\n      ".$self->url();

    return $s;
}

1;

__END__

=head1 NAME

C<Net::TiVo::Show> - Class that wraps the XML interface that defines a
TiVo show.

=head1 SYNOPSIS

    use Net::TiVo;
	
    my $tivo = Net::TiVo->new(
        host => '192.168.1.25', 
        mac  => 'MEDIA_ACCESS_KEY'
    );

    for my $folder ($tivo->folders()) {
        for my $show ($folder->shows()) {
            print $show->as_string(), "\n";
        }
    }

=head1 DESCRPTION

C<Net::TiVo::Show> provides an object-oriented interface to an XML description
of a TiVo show.  It provides the necessary accessors to read the XML data. 

=head2 METHODS

=over 4

=item station_name()

Returns TiVo's name for this station, i.e. FoodTV is FOOD.

=item name()

Returns the name of this show.

=item episode()

Returns the title of this show.

=item episode_num()

Returns the episode number of the show.

=item content_type()

Returns the content type of this show in MIME format.

=item capture_date()

Returns the date this show was recorded in seconds since the epoch.

=item format()

Returns the source format of this show.

=item high_definition()

Returns Yes if the show was recorded in high definition, or No if it was not.

=item size()

Returns the size of this show in bytes.

=item channel()

Returns the channel this show was recorded on.

=item tuner()

Returns the number of tuner used to record the show.

=item duration()

Returns the duration of this show in milliseconds.

=item description()

Returns the description of this show.

=item program_id()

Returns a hexadecimal string containing the program id.

=item series_id()

Returns a hexadecimal string containing the series id.

=item url()

Returns the url of this show.  This information can be used to download the
episode from TiVo to machine.

=item as_string()

Returns a pretty print of the this show's information, including title, episode
title, description, duration in minutes, and url.

=back

=head1 SEE ALSO

L<Net::TiVo>, L<Net::TiVo::Folder>

=head1 AUTHOR

Christopher Boumenot, E<lt>boumenot@gmail.comE<gt>

=cut


lib/Net/TiVo/Folder.pm
# $Id: Folder.pm 57 2007-01-12 19:26:09Z boumenot $
# Author: Christopher Boumenot <boumenot@gmail.com>
######################################################################
#
# Copyright 2006-2007 by Christopher Boumenot.  This program is free 
# software; you can redistribute it and/or modify it under the same
# terms as Perl itself.
#
######################################################################

package Net::TiVo::Folder;

use strict;
use warnings;

use Text::Wrap;
use Log::Log4perl qw(:easy get_logger);

# Should be read as a poor man's XPath
our %DEFAULT_ATTRIBUTES_XPATH = (
    content_type => [qw(Details ContentType)],
    format       => [qw(Details SourceFormat)],
    change_date  => [qw(Details LastChangeDate)],
    name         => [qw(Details Title)],
    total_items  => [qw(Details TotalItems)],
    # Is this the same thing as total_items()?
    item_count   => [qw(ItemCount)],
    # What value is this?
    item_start   => [qw(ItemStart)],
    global_sort  => [qw(GlobalSort)],
    sort_order   => [qw(SortOrder)],
    # Due to the way folders are created it is difficult
    # to get the url of the folder.  I could add it easily
    # enough, but it would be a hack, and I'm not sure 
    # that you really need the url of the folder.  I need
    # to fix the code.
    #url
);

__PACKAGE__->make_accessor($_) for keys %DEFAULT_ATTRIBUTES_XPATH;
__PACKAGE__->make_accessor($_) for qw(size);
__PACKAGE__->make_array_accessor($_) for qw(shows);

sub TIVO_MIME_TYPES { qw(video/x-tivo-mpeg video/x-tivo-raw-pes) }

sub new {
    my ($class, %options) = @_;
    
    unless ($options{xmlref}) {
        die __PACKAGE__ . ": Mandatory param xmlref missing\n";
    }

    my $self = {
        %options,
    };

    bless $self, $class;

    for my $attr (keys %DEFAULT_ATTRIBUTES_XPATH) {
        my $value = __PACKAGE__->walk_hash_ref($options{xmlref}, $DEFAULT_ATTRIBUTES_XPATH{$attr});
        $self->$attr($value);
    }

    $self->change_date(hex($self->change_date()));

#    DEBUG(sub { Data::Dumper::Dumper($options{xmlref}) });

    my ($size, @shows);
    for my $show (@{$options{xmlref}->{Item}}) {
        if (grep {$show->{Links}->{Content}->{ContentType} eq $_} TIVO_MIME_TYPES) {
            push @shows, Net::TiVo::Show->new(xmlref => $show);
            INFO("added the show " . $shows[-1]->name());
            $size += $shows[-1]->size();
        } 
    }

    $self->size($size);
    $self->shows(\@shows);
    
    return $self;
}

sub _commify {
    my ($self, $num) = @_;
    $num = reverse $num;
    $num =~ s<(\d\d\d)(?=\d)(?!\d*\.)><$1,>g;
    $num = reverse $num;
    return $num;
}

sub as_string {
    my $self = shift;

    $Text::Wrap::columns = 72;

    my @a;
    push @a, $self->name();
    push @a, sprintf("%d %s", $self->total_items(), (($self->total_items() > 1) ? "episodes" : "episode"));
    push @a, sprintf("%s bytes", $self->_commify($self->size()));
    my $s = wrap("", "      ", join(", ", @a));
    return $s;
}

# cmb - taken from the excellent Net::Amazon!
sub make_accessor {
    my($package, $name) = @_;

    no strict qw(refs);

    my $code = <<EOT;
    *{"$package\\::$name"} = sub {
        my(\$self, \$value) = \@_;

        if(defined \$value) {
            \$self->{$name} = \$value;
        }
        if(exists \$self->{$name}) {
            return (\$self->{$name});
        } else {
            return "";
        }
    }
EOT
    if(! defined *{"$package\::$name"}) {
        eval $code or die "$@";
    }
}

# cmb - taken from the excellent Net::Amazon!
sub make_array_accessor {
    my($package, $name) = @_;

    no strict qw(refs);

    my $code = <<EOT;
    *{"$package\\::$name"} = sub {
        my(\$self, \$nameref) = \@_;
        if(defined \$nameref) {
            if(ref \$nameref eq "ARRAY") {
                \$self->{$name} = \$nameref;
            } else {
                \$self->{$name} = [\$nameref];
            }
        }
        # Return a list
        if(exists \$self->{$name} and
           ref \$self->{$name} eq "ARRAY") {
            return \@{\$self->{$name}};
        }

        return undef;
    }
EOT

    if(! defined *{"$package\::$name"}) {
        eval $code or die "$@";
    }
}

sub walk_hash_ref {
    my ($package, $href, $aref) = @_;

    return $href if scalar(@$aref) == 0;

    my @a;
    push @a, $_ for @$aref;

    my $tail = pop @a;
    my $ref = $href;

    for my $part (@a) {
        $ref = $ref->{$part};
    }
    
    return $ref->{$tail};
}

1;

__END__

=head1 NAME

Net::TiVo::Folder - Class that wraps the XML description that defines a TiVo
folder.

=head1 SYNOPSIS

    use Net::TiVo;
	
    my $tivo = Net::TiVo->new(
        host => '192.168.1.25', 
        mac  => 'MEDIA_ACCESS_KEY'
    );
        
    for my $folder ($tivo->folders()) {
        print $folder->as_string(), "\n";
    }

=head1 DESCRPTION

C<Net::TiVo::Folder> provides an object-oriented interface to an XML
description of a TiVo show.  It provides the necessary accessors to read
the XML data.

=head2 METHODS

=over 4

=item content_type()

Returns TiVo's mime type for a folder (x-tivo-container/tivo-videos).

=item format()

Returns TiVo's mime type for the format of the folder
(x-tivo-container/tivo-dvr).

=item change_date()

Returns the last time the folder was changed.  The value is in seconds
since the epoch.

=item name()

Returns the name of the folder.

=item total_items()

Returns the number of shows contained in this folder.

=item global_sort()

Returns a boolean (Yes or No) indicating if the folder is globally
sorted.

=item sort_order()

Returns the sort order of the folder.

=item size()

Returns the size in bytes of this folder.  This value is calculated by
summing the individual shows contained in this folder.

=item shows()

Returns an array of the shows contained in this folder.

=item as_string()

Returns a pretty print of this folder's information, including name,
number of show, size, and a list of shows contained in the folder.

=back

=head1 SEE ALSO

L<Net::TiVo>, L<Net::TiVo::Show>

=head1 AUTHOR

Christopher Boumenot, E<lt>boumenot@gmail.comE<gt>

=cut


lib/Net/TiVo.pm
# $Id: TiVo.pm 63 2007-03-29 14:09:37Z boumenot $
# Author: Christopher Boumenot <boumenot@gmail.com>
######################################################################
#
# Copyright 2006-2007 by Christopher Boumenot.  This program is free 
# software; you can redistribute it and/or modify it under the same
# terms as Perl itself.
#
######################################################################

package Net::TiVo;

use strict;
use warnings;

our $VERSION = '0.10';

use LWP::UserAgent;
use HTTP::Request;
use XML::Simple;
use Data::Dumper;
use Log::Log4perl qw(:easy get_logger);
use Net::TiVo::Folder;
use Net::TiVo::Show;

use constant TIVO_URL => '/TiVoConnect?Command=QueryContainer&Container=%2FNowPlaying';

sub new { 
    my $class = shift;
    my $self = {username => 'tivo',
                realm    => 'TiVo DVR',
                @_};

    $self->{host}     || die "%Error: no host was defined!\n";
    $self->{mac}      || die "%Error: no mac was defined!\n";
    $self->{username} || die "%Error: no username was defined!\n";

    $self->{ua}  = LWP::UserAgent->new() or
        die "%Error: failed to create a LWP::UserAgent!";

    $self->{ua}->credentials($self->{host}.':443',
                             $self->{realm}, 
                             $self->{username} => $self->{mac});

    $self->{url} = 'https://'.$self->{host}.TIVO_URL;

    bless $self, $class;
    return $self;
}


sub folders {
    my $self = shift;

    my $resp = $self->_fetch($self->{url});

    if ($resp->is_success()) {
        my @folders;
        $self->_parse_content($resp->content(), \@folders);

        return unless @folders;
        return (wantarray) ? @folders : \@folders;
    } 

    print "%Error: $resp->status_line()!\n";
    return;
}


sub _fetch {
    my ($self, $url) = @_;
    my $resp;

    INFO("fetching $url");

    if (exists $self->{cache}) {
        $resp = $self->{cache}->get($url, $resp);
        if (defined $resp) {
            INFO("cache hit");
            return $resp;
        }
        INFO("cache miss");
    } 

    $resp = $self->{ua}->request(HTTP::Request->new(GET => $url));
    die "%Error: fetch failed, " . $resp->status_line() . "!\n" unless $resp->is_success();

    if (exists $self->{cache}) {
        $self->{cache}->set($url, $resp);
    }

    return $resp;
}

sub _parse_content {
    my ($self, $cont, $folder_aref) = @_;

    DEBUG(sub { "Received [" . $cont . "]"});

    my $xmlref = XMLin($cont, ForceArray => ['Item']);
    unless (defined $xmlref->{Item}) {
        INFO("No content to parse, skipping ...");
        return;
    }
    
    DEBUG(sub { Dumper($xmlref) });

    # TiVo only allows you to create one folder to hold shows, but the
    # top most folder, Now Playing, as to be accounted for too.  If we
    # haven't created any folders yet, then this is the Now Playing
    # folder, and needs to be treated specially.
    push @$folder_aref, Net::TiVo::Folder->new(xmlref => $xmlref);

    # 2006/12/29 - RHARMAN: TiVo Suggestions can exist but contain zero videos
    if ($folder_aref->[-1]->total_items() > 0) {
        INFO("added the folder " . $folder_aref->[-1]->name());
    } else {
        INFO("skipped the folder " , $folder_aref->[-1]->name(), " because it was empty.");
        pop @$folder_aref;
    }

    for my $i (@{$xmlref->{Item}}) {
        my $ct = $i->{Links}->{Content}->{ContentType};


        if ($ct eq 'x-tivo-container/folder') {
            my $resp = $self->_fetch($i->{Links}->{Content}->{Url});
            $self->_parse_content($resp->content(), $folder_aref);
        } else {
            INFO("skipping the content for $ct");
        }
    }
}


1;
__END__

=head1 NAME

Net::TiVo - Perl interface to TiVo.

=head1 SYNOPSIS

    use Net::TiVo;

    my $tivo = Net::TiVo->new(
        host => '192.168.1.25', 
        mac  => 'MEDIA_ACCESS_KEY'
    );

    for ($tivo->folders()) {
        print $_->as_string(), "\n";
    }	

=head1 ABSTRACT

C<Net::TiVo> provides an object-oriented interface to TiVo's REST interface.
This makes it possible to enumerate the folders and shows, and dump their
meta-data.

=head1 DESCRIPTION

C<Net::TiVo> has a very simple interface, and currently only supports the
enumeration of folder and shows using the REST interface.  The main purpose of
this module was to provide access to the TiVo programmatically to automate the
process of downloading shows from a TiVo.

C<Net::TiVo> does not provide support for downloading from TiVo.  There are
several options available, including LWP, wget, and curl.  Note: I have used
wget version >= 1.10 with success.  wget version 1.09 appeared to have an issue
with TiVo's cookie.

=head1 BUGS

One user has reported 500 errors when using the library.  He was able to track
the bug down to LWP and Net::SSLeay.  Once he switched from using Net::SSLeay
to Crypt::SSLeay the 500 errors went away.

=head1 CACHING

C<Net::TiVo> is slow due to the amount of time it takes to fetch data from
TiVo.  This is greatly sped up by using a cache.  C<Net::TiVo>'s C<new> method
accepts a reference to a C<Cache> object.  Any type of caching object may be
supported as long as it meets the requirements below.  There are several cache
implementations available on CPAN, such as C<Cache::Cache>.

The following example creates a cache that lasts for 600 seconds.

    use Cache::FileCache;
    
    my $cache = Cache::FileCache->new(
         namespace          => 'TiVo',
         default_expires_in => 600,
    }

    my $tivo = Net::TiVo->new(
         host  => '192.168.1.25',
         mac   => 'MEDIA_ACCESS_KEY',
         cache => $cache,
    }

C<Net::TiVo> uses I<positive> caching, errors are not stored in the cache.

Any C<Cache> class may be used as long as it supports the following method
signatures.

    # Set a cache value
    $cache->set($key, $value);

    # Get a cache value
    $cache->get($key);


=head2 METHODS

=over 4

=item folders()

Returns an array in list context or array reference in scalar context
containing a list of Net::TiVo::Folder objects.

=back

=head1 SEE ALSO

L<Net::TiVo::Folder>, L<Net::TiVo::Show>

=head1 AUTHOR

Christopher Boumenot, E<lt>boumenot@gmail.comE<gt>

=cut
